### DISCUSSION TITLE
General question about the String class and memory management

### DISCUSSION THREAD

**Discussion by mjbmikeb2** [electric-sheep-co/arduino-redis#53] [General]
As far as I can tell, the library makes extensive use of the Arduino String class which is notorious for causing heap fragmentation on microcontrollers. The general advice from Arduino experts is "don't use it, use char arrays instead".

I went looking for an alternative non-fragmenting String class but couldn't find a suitable drop-in replacement with the same interface and behavior.

Looking at the other code it looks like the project started life on a standard computer with a proper MMU rather than a microcontroller with a simplified MMU. Is this is a correct observation? If so, has there been any attempt to optimize the code for memory limited environment of a microcontroller?

I would really like to use this library, but I don't want to invest time in something if several weeks down the road I start running into memory fragmentation issues for which there is no workaround other than a periodic reboot.

**Comment by rpj** [comment]
Hi @mjbmikeb2, thanks a ton for reaching out and your interest!

> I went looking for an alternative non-fragmenting String class but couldn't find a suitable drop-in replacement with the same interface and behavior.

I had done a very quick search for something similar a couple months after putting together the initial version of this library and also couldn't find anything suitable, and with limited time just gave up on that search.

> Looking at the other code it looks like the project started life on a standard computer with a proper MMU rather than a microcontroller with a simplified MMU. Is this is a correct observation? If so, has there been any attempt to optimize the code for memory limited environment of a microcontroller?

Yup that's basically correct coupled with the fact that while I'm a seasoned software developer I was an Arduino neophyte upon starting this library and simply wasn't aware of `String`'s issues. As noted above I eventually found out about them too and went looking a bit for a replacement but didn't find anything usable. So unfortunately there's no latent branch with some half-baked work in this regard. 

I have added this as a task in an issue that was already tracking general memory model cleanup, #8, and would heartily welcome any contributions in this regard!

(Apologies for initially replying from a different account, I'm always getting them mixed up 🤦 )
